zedUpload/azureutil/azure.go

```diff
@@ -312,7 +312,20 @@ func DownloadAzureBlob(accountURL, accountName, accountKey, containerName, remot
 			if err != nil {
 				return err
 			}
-			body := dr.Body(azblob.RetryReaderOptions{MaxRetryRequests: maxRetries})
+			var retryErrorStr []string
+			body := dr.Body(azblob.RetryReaderOptions{
+				MaxRetryRequests: maxRetries,
+				NotifyFailedRead: func(failureCount int, lastError error, offset int64, count int64, willRetry bool) {
+					retryError := fmt.Errorf("total read failures: %d, offset: %d, error: %v", failureCount, offset, lastError)
+					retryErrorStr = append(retryErrorStr, retryError.Error())
+					if !willRetry {
+						err = fmt.Errorf(strings.Join(retryErrorStr, "\n"))
+					}
+				},
+			})
+			if err != nil {
```

Comment:
I think this `if err != nil {` is not needed. As I understand it, `dr.Body` is just constructor and download starts below with `io.CopyBuffer`.

Reply:
Also, is this async in some way? I haven't looked at the azure API in quite some time.

Reply:
I do not know about the whole API, but the way we use it is as follows: We send 16 concurrent requests to download 1MB chunks. When we download all of them, we move to the next 16 chunks until we download the whole file.